Addressing port/endpoint issues on Qdrant

timheuer opened the pull request:
Fix #3421
- Separate port params
- update playground
- Modify test to verify port in manifest
- Separated EndpointReferences
 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/3422)

timheuer committed:
Addressing port/endpoint issues on Qdrant

davidfowl commented on src/Aspire.Hosting.Qdrant/QdrantBuilderExtensions.cs:
```diff
@@ -27,20 +27,22 @@ public static class QdrantBuilderExtensions
         string name,
         IResourceBuilder<ParameterResource>? apiKey = null,
-        int? port = null)
+        int? gRpcPort = null,
```
```suggestion
        int? grpcPort = null,
```

davidfowl commented on playground/Qdrant/Qdrant.AppHost/Program.cs:
```diff
@@ -3,7 +3,7 @@
 var builder = DistributedApplication.CreateBuilder(args);
 
-var qdrant = builder.AddQdrant("qdrant")
+var qdrant = builder.AddQdrant("qdrant", httpPort:5503)
```
Remove the port?

davidfowl commented on playground/Qdrant/Qdrant.AppHost/Program.cs:
```diff
@@ -3,7 +3,7 @@
 var builder = DistributedApplication.CreateBuilder(args);
 
-var qdrant = builder.AddQdrant("qdrant")
+var qdrant = builder.AddQdrant("qdrant", httpPort:5503)
```
```suggestion
var qdrant = builder.AddQdrant("qdrant")
```

remove

timheuer replied:
Was in there to have a validated manifest test...you don't see the need?

timheuer committed:
PR feedback (param name)

eerhardt reviewed (approved):
LGTM

eerhardt commented on tests/Aspire.Hosting.Tests/Qdrant/AddQdrantTests.cs:
```diff
@@ -174,7 +174,7 @@ public async Task QdrantClientAppWithReferenceContainsConnectionStrings()
     {
         var appBuilder = DistributedApplication.CreateBuilder(new DistributedApplicationOptions() { Args = new string[] { "--publisher", "manifest" } } );
-        var qdrant = appBuilder.AddQdrant("qdrant");
+        var qdrant = appBuilder.AddQdrant("qdrant", httpPort:5503);
```
Instead of modifying this test, we should add a new one. I can push a new test to this branch.

eerhardt committed:
Fix up tests

eerhardt commented on src/Aspire.Hosting.Qdrant/QdrantBuilderExtensions.cs:
```diff
@@ -27,20 +27,22 @@ public static class QdrantBuilderExtensions
             .WithImage(QdrantContainerImageTags.Image, QdrantContainerImageTags.Tag)
-            .WithHttpEndpoint(port: port, targetPort: QdrantPortGrpc, name: QdrantServerResource.PrimaryEndpointName)
-            .WithHttpEndpoint(port: port, targetPort: QdrantPortHttp, name: QdrantServerResource.RestEndpointName)
+            .WithHttpEndpoint(port: grpcPort, targetPort: QdrantPortGrpc, name: QdrantServerResource.PrimaryEndpointName)
+            .WithHttpEndpoint(port: httpPort, targetPort: QdrantPortHttp, name: QdrantServerResource.RestEndpointName)
```
Is it confusing that the `httpPort` goes to the `rest` endpoint, and not the endpoint that is named `http`?

Maybe we should either rename the endpoints, or rename these port parameters?

timheuer replied:
I thought about this as developing. I was trying to match the way the Qdrant docs refer to the exposed API endpoints (as gRPC and REST).  But the more I think about it, it's mixing weird names.  

As you re-raise the question here @eerhardt , I think maybe representing the endpoints what they are (grpc and http) eliminating the 'high level' speak is maybe better...naming them clearly for what the endpoint communicates.

Thoughts? Broader quick rename -- change default endpoint name from 'http' to 'grpc' and change 'rest' to 'http' everywhere?

eerhardt replied:
I changed the default endpoint name to 'grpc' but left the other one 'rest`. I can rename it to 'http' everywhere if you'd like.

The part that I don't like about using 'http' is that they are both HTTP. Where at least "rest" gives a little more indication of the difference. But I can rename it to 'http' if you'd like.

timheuer replied:
or the way you just did ;-)

eerhardt replied:
Looking for a little bit of previous art, Dapr says:

https://github.com/dotnet/aspire/blob/e2b1b46044bb74e9d7ccce54d00ada58d411b947/src/Aspire.Hosting.Dapr/DaprDistributedApplicationLifecycleHook.cs#L154-L156

Not saying we'd have to match it, but it at least shows we use this naming pattern elsewhere.

eerhardt replied:
I was thinking Qdrant docs being the winner. 😆 

I can honestly go either way. If I had to pick I would pick the way it is now in the PR mostly because it matches how Qdrant refers to them. But I can live with renaming "rest" => "http" if you'd like.

timheuer replied:
Ugh, sorry not having this convo in original PR.  Let's rename to http -- the docs are mixed. But the **config** docs consistently use 'http_port' as what is being exposed, let's rename rest->http

eerhardt committed:
Respond to PR feedback.

Rename ports and endpoint names to be consistent.

timheuer committed:
PR feedback
- moved to using grpc/http endpoint name consistent with config and prior art
- fixed wrong link in readme
- fixed/validated resource hosting test
